File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/replay/ReplayRecordings.kt
package expo.modules.vescapecore.replay

import expo.modules.vescapecore.recording.DebugRecordingStore

import android.content.Context
import org.json.JSONObject
import java.io.File

/**
 * Resolves a replayable Debug Recording by name: the on-device store first, then the bundled
 * fixtures shipped in app assets (`assets/fixtures`, copied from `shared/fixtures/` by
 * `copy:shared`). Bundled fixtures make replay usable on a device with no captures yet.
 *
 * @parity /modules/vescape-core/ios/replay/ReplayRecordings.swift
 */
internal object ReplayRecordings {
    private const val ASSETS_DIR = "fixtures"

    fun read(context: Context, name: String): String {
        val store = DebugRecordingStore(context)
        if (store.exists(name)) return store.read(name)
        requireValidName(name)
        return context.assets.open("$ASSETS_DIR/$name").bufferedReader().readText()
    }

    /** The recording's `meta` first line, or null when missing/malformed (truncated capture). */
    fun readMeta(context: Context, name: String): JSONObject? =
        read(context, name).lineSequence().firstOrNull()?.let { line ->
            try {
                JSONObject(line).takeIf { it.optString("kind") == "meta" }
            } catch (e: Exception) {
                null
            }
        }

    /** Bundled fixture names + sizes, sorted by name (assets have no timestamps). */
    fun listBundled(context: Context): List<Map<String, Any>> =
        (context.assets.list(ASSETS_DIR) ?: emptyArray())
            .filter { it.endsWith(".jsonl") }
            .sorted()
            .map { name ->
                val sizeBytes = context.assets.open("$ASSETS_DIR/$name").use { stream ->
                    var total = 0L
                    val buffer = ByteArray(8 * 1024)
                    while (true) {
                        val read = stream.read(buffer)
                        if (read < 0) break
                        total += read
                    }
                    total
                }
                mapOf("name" to name, "sizeBytes" to sizeBytes)
            }

    private fun requireValidName(name: String) {
        require(File(name).name == name && name.endsWith(".jsonl")) { "Invalid debug recording name" }
    }
}

File: modules/vescape-core/ios/replay/ReplayRecordings.swift
import Foundation

/// Resolves a replayable Debug Recording by name: the on-device store first, then the bundled
/// fixtures shipped in the `VescapeCoreAssets` resource bundle (symlinked from `shared/fixtures/`
/// via the podspec). Bundled fixtures make replay usable on a device with no captures yet.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/replay/ReplayRecordings.kt
internal enum ReplayRecordings {
  /// Resolve a recording to a readable URL: on-device store first, bundled fixture second.
  static func url(name: String) -> URL? {
    if let stored = DebugRecordingStore.recordingURL(name: name) { return stored }
    guard (name as NSString).lastPathComponent == name, name.hasSuffix(".jsonl") else { return nil }
    return bundledFixtureURLs().first { $0.lastPathComponent == name }
  }

  /// Bundled fixture names + sizes, sorted by name (bundle resources have no capture timestamps).
  static func listBundled() -> [[String: Any]] {
    bundledFixtureURLs()
      .sorted { $0.lastPathComponent < $1.lastPathComponent }
      .map { url in
        let sizeBytes = Int64((try? url.resourceValues(forKeys: [.fileSizeKey]))?.fileSize ?? 0)
        return ["name": url.lastPathComponent, "sizeBytes": sizeBytes]
      }
  }

  /// Fixture `.jsonl` files, whether they landed in the CocoaPods resource bundle
  /// (`VescapeCoreAssets`) or directly in the module bundle (same lookup as `cell-presets.json`).
  private static func bundledFixtureURLs() -> [URL] {
    let moduleBundle = Bundle(for: DebugRecordingStore.self)
    let bundles = [
      moduleBundle,
      moduleBundle.url(forResource: "VescapeCoreAssets", withExtension: "bundle").flatMap { Bundle(url: $0) },
    ]
    for case let bundle? in bundles {
      let urls = bundle.urls(forResourcesWithExtension: "jsonl", subdirectory: nil) ?? []
      if !urls.isEmpty { return urls }
    }
    return []
  }
}
